cep-recipe-jinja.md

```diff
@@ -0,0 +1,125 @@
+# Jinja functions in recipes
+
+The new recipe format has some Jinja functionalities. We want to specify what functions exist and their expected behavior.
+
+## The compiler function
+
+The compiler function is used to stick together a compiler from {lang}_compiler and {lang}_compiler_version
+
+The function looks as follows:
+
+```yaml
+${{ compiler('c') }}
+```
+
+This would pull in the c_compiler and c_compiler_version from the variant config. The compiler function also adds the target_platform to render to something such as:
+
+```
+gcc_linux-64 8.9
+clang_osx-arm64 12
+msvc_win-64 19.29
+```
+
+The function thus evaluates to `{compiler}_{target_platform} {compiler_version}`.
+
+The variant config could look something like:
+
+```
+c_compiler:
+  - gcc
+c_compiler_version:
+  - "8.9"
+cxx_compiler:
+  - clang
+cxx_compiler_version:
+  - "12"
+```
+
+## The `cdt` function
+
+cdt stands for "core dependency tree" packages. These are repackaged from Centos.
+
+The function expands to the following: 
+
+- package-name-<cdt_name>-<cdt_arch>
+
+Where `cdt_name` and `cdt_arch` are loaded from the variant config. If they are undefined, they default to:
+
+- `cos6` for `cdt_name` on `x86_64` and `x86`, otherwise `cos7`
+- To the `platform::arch` for `cdt_arch`, except for `x86` where it defaults to `i686`.
+
+## The `pin_subpackage` function
+
+### Pin definition
+
+A pin has the following arguments:
+
+- `min_pin`: The lower bound of the dependency spec. This is expressed as a `x.x....` version where the `x` are filled in from the corresponding version of the package. For example, `x.x` would be `1.2` for a package version `1.2.3`. The resulting pin spec would look like `>=1.2` for the `min_pin` argument of `x.x` and a version of `1.2.3` for the package.
+- `max_pin`: This defines the upper bound and follows the same `x.x` semantics but adds `+1` to the last segment. For example, `x.x` would be `1.(2 + 1)` for a package version `1.2.3`. The resulting pin spec would look like `<1.3` for the `max_pin` argument of `x.x` and a version of `1.2.3` for the package.
```

Comment:
There seems a strong but unstated assumption here that version strings are integers separated by dots (`.`). If that is the case, that assumption should be explicitly stated and guidance provided as to what should happen if the assumption is violated. (Just sayin', because `1!24.4.0+91_gbfa8ccdce` is a perfectly valid PEP-440/conda package version string.)

Reply:
Also, as written, it seems like `min_pin` and `max_pin` do not support an explicit version; e.g., recipes cannot simply have `max_pin="1.2.3"`.

Reply:
> Also, as written, it seems like `min_pin` and `max_pin` do not support an explicit version; e.g., recipes cannot simply have `max_pin="1.2.3"`.

Conda's existing `pin_compatible` jinja [has](https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html#extra-jinja2-functions) _both_ `{min,max}_pin` as well as `{lower,upper}_bound`, in order to be able to distinguish `'x.x'` from an explicit version limit.

Perhaps it's easier to understand to overload both styles into two kwargs (one for lower, one for upper), and allow either `'x(.x)*'` or an explicit version number? It's more work to implement, but less confusing IMO (otherwise it's easy to specify conflicting things e.g. by setting both `max_pin` and `upper_bound`)

Reply:
Wow, I wasn't aware! I found a few places where it's used but I don't know how it works. For example, in the `onnx-feedstock` it says:

```
    - {{ pin_compatible('numpy', lower_bound='1.19', upper_bound='3.0') }}  # [py>38]
```

Does this still take the lower bound from the build-time resolution? Or is this equivalent to just writing `numpy >=1.19,<3.0`?

Reply:
I also don't really understand why we would want this in the function. One can just as well write something like:

```
- numpy >=1.5
- ${{ pin_compatible("numpy", max_pin="x.x", min_pin="x.x.x") }}
```

If one needs additional (hard-coded) constraints. The solver doesn't really care about more constraints.

Reply:
> I also don't really understand why we would want this in the function.

I wanted this for numpy2, but that's a rare enough case that it probably doesn't matter. In that case, we're building against 2.0, but I want the lower bound to be 1.19, which cannot be done with what you propose (because it's _looser_ than what `pin_compatible` enforces).

More generally though, in your example, I find that adding additional constraints just feels like a weaksauce work-around. If I'm already using a function to specify the constraint on numpy, why won't it give me the expressivity needed to do that?

> Does this still take the lower bound from the build-time resolution? Or is this equivalent to just writing `numpy >=1.19,<3.0`?

I think the intention of the author was clear (be compatible with numpy 2 on py>38), but it doesn't work because it just creates an _additional_ constraint, where - again - it would need to be looser. IOW, onnx's [metadata](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Flinux-64%2Fonnx-1.16.0-py39h974b0fd_0.conda) contains both:
```
numpy >=1.19,<3.0
numpy >=1.26.4,<2.0a0
```

Reply:
> but that's a rare enough case that it probably doesn't matter

On second thought, probably it does. That onnx recipe is precisely trying to work around the same thing that I need to handle in the numpy run-export - in other words, building against numpy >=1.25 (with no extra manual overrides) gives a run-time lower bound of `>= 1.19` currently.

Reply:
> but it doesn't work because it just creates an _additional_ constraint, where - again - it would need to be looser.

Ah wait, maybe it actually does work correctly, and the package just falls prey to the numpy run-export, which adds the `numpy >=1.26.4,<2.0a0` on top`.

Reply:
@wolfv @chenghlee WDYT about the idea of having pin-or-version arguments in `pin_compatible`? As an example, let's say for numpy 2.0.1:
``` 
pin_compatible("numpy")                                         --> numpy >=2.0.1,<3    # default: lower_bound='x.x.x.x.x.x', upper_bound='x'
pin_compatible("numpy", upper_bound="x.x")                      --> numpy >=2.0.1,<2.1
pin_compatible("numpy", lower_bound="x.x", upper_bound="2.3")   --> numpy >=2.0,<2.3
pin_compatible("numpy", lower_bound="1.19")                     --> numpy >=1.19,<3     # most libs compiled against np2
pin_compatible("numpy", lower_bound="1.19", upper_bound="2.3")  --> numpy >=1.19,<2.3   # scipy's usecase
```

The API contract could be that anything matching `x(\.x)*` is a pinning expression, and the rest gets treated as an explicit version.

Reply:
@wolfv, given your changes in 0fda2133b485275ae0ef59945e22afd61473937e, I presume you're not on board with this idea (or didn't see this thread)?

If you want to keep `{min,max}_pin` separate from `{lower,upper}_bound`, I think we need to enforce that at most one `{min_pin,lower_bound}` resp. `{max_pin,upper_bound}` each may be specified, everything else should raise an error.

I'd still prefer a unified interface, because I don't see the gain to keep these arguments separate (but rather requires a lot of complexity to ensure it is unambiguous/consistent)

Reply:
I indeed didn't see it or didn't read carefully enough. I also think this is a good change. The logic is already in place, so I don't think it will be a big change on the rattler-build side.

Reply:
I proposed the API that I think is more natural and IMO more desirable long-term.

Obviously the naming choice is just bike shedding and both would work, but `max_pin=1.2.3` seems off to me, whereas `upper_bound='x.x'` is fine because it's a more generic term, where users don't have to guess whar a pin is or isn't.